### ISSUE TITLE
loss.backward(): Unsupported dimension, only the last dimension can differ

### ISSUE THREAD

**Issue by awoimbee** [tracel-ai/burn#3427]
**Describe the bug**

I previously opened an issue that I solved by using the NdArray backend (#3420).
But for this one idk what to do.

```
thread 'main' panicked at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-ndarray-0.18.0/src/ops/base.rs:645:17:
Unsupported dimension, only the last dimension can differ: Tensor [1, 1000, 1000] Index [1, 1, 1000]
stack backtrace:
   0: __rustc::rust_begin_unwind
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/std/src/panicking.rs:697:5
   1: core::panicking::panic_fmt
             at /rustc/6b00bc3880198600130e1cf62b8f8a93494488cc/library/core/src/panicking.rs:75:14
   2: burn_ndarray::ops::base::NdArrayMathOps<E>::gather_batch_size
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-ndarray-0.18.0/src/ops/base.rs:645:17
   3: burn_ndarray::ops::base::NdArrayMathOps<E>::scatter
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-ndarray-0.18.0/src/ops/base.rs:574:26
   4: burn_ndarray::ops::tensor::<impl burn_tensor::tensor::ops::tensor::FloatTensorOps<burn_ndarray::backend::NdArray<E,I,Q>> for burn_ndarray::backend::NdArray<E,I,Q>>::float_scatter::{{closure}}
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-ndarray-0.18.0/src/ops/tensor.rs:165:68
   5: burn_ndarray::ops::tensor::<impl burn_tensor::tensor::ops::tensor::FloatTensorOps<burn_ndarray::backend::NdArray<E,I,Q>> for burn_ndarray::backend::NdArray<E,I,Q>>::float_scatter
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-ndarray-0.18.0/src/tensor.rs:100:49
   6: <burn_autodiff::ops::maxmin::MaxMinDim as burn_autodiff::ops::backward::Backward<B,1_usize>>::backward::{{closure}}
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/ops/maxmin.rs:23:13
   7: burn_autodiff::ops::backward::unary
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/ops/backward.rs:85:20
   8: <burn_autodiff::ops::maxmin::MaxMinDim as burn_autodiff::ops::backward::Backward<B,1_usize>>::backward
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/ops/maxmin.rs:17:9
   9: <burn_autodiff::ops::base::OpsStep<B,T,SB,_> as burn_autodiff::graph::base::Step>::step
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/ops/base.rs:257:9
  10: burn_autodiff::runtime::server::AutodiffServer::execute_steps::{{closure}}::{{closure}}
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/server.rs:99:34
  11: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:798:29
  12: <alloc::vec::into_iter::IntoIter<T,A> as core::iter::traits::iterator::Iterator>::fold
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/into_iter.rs:323:25
  13: core::iter::traits::iterator::Iterator::for_each
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:801:9
  14: burn_autodiff::runtime::server::AutodiffServer::execute_steps::{{closure}}
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/server.rs:99:18
  15: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:798:29
  16: core::iter::traits::double_ended::DoubleEndedIterator::rfold
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/double_ended.rs:308:21
  17: <core::iter::adapters::rev::Rev<I> as core::iter::traits::iterator::Iterator>::fold
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/rev.rs:64:9
  18: core::iter::traits::iterator::Iterator::for_each
             at /home/arthur/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:801:14
  19: burn_autodiff::runtime::server::AutodiffServer::execute_steps
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/server.rs:96:32
  20: burn_autodiff::runtime::server::AutodiffServer::backward
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/server.rs:42:25
  21: <burn_autodiff::runtime::mutex::MutexClient as burn_autodiff::runtime::client::AutodiffClient>::backward
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/runtime/mutex.rs:40:20
  22: <burn_autodiff::backend::Autodiff<B,C> as burn_tensor::tensor::backend::base::AutodiffBackend>::backward
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-autodiff-0.18.0/src/backend.rs:63:9
  23: burn_tensor::tensor::api::autodiff::<impl burn_tensor::tensor::api::base::Tensor<B,_>>::backward
             at /home/arthur/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-tensor-0.18.0/src/tensor/api/autodiff.rs:8:9
  24: <burn_mnist::model::GeometryAutoEncoder<B> as burn_train::learner::train_val::TrainStep<burn_mnist::data::PointCloudBatch<B>,burn_train::learner::regression::RegressionOutput<B>>>::step
             at ./src/model/mod.rs:56:32
  25: burn_train::learner::epoch::TrainEpoch<B,TI>::run
```

1_000 is the number of points in my point cloud.

**To Reproduce**

```bash
git clone https://github.com/awoimbee/burn-geometry-experiment
cd burn-geometry-experiment
git checkout fc06a11ffcdf3ccdecfdbf7a0f2694cd9e4b2b7f
# download an stl file, like https://www.thingiverse.com/thing:4564567
mkdir -p dataset/train
mkdir -p dataset/test
mv <the-stl-file> dataset/train/geom.stl
cp dataset/train/geom.stl dataset/test/geom.stl
cargo run -- train
```

**Expected behavior**

Training works.

**Desktop (please complete the following information):**
 - OS: linux 6.15
 - Backend: NdArray, WGPU (vulkan)

**Comment by nathanielsimard** [comment]
Maybe that's a broadcasting issue in the backward pass for a specific operation. If you can pinpoint the operation we could easily make a fix.

**Comment by awoimbee** [comment]
I commented out most of my code, looks like the issue is `max_dim(2)` ?

```rust
        // 1. k-NN indices [1, 1000, 32]
        let idx = batched_knn(points.clone(), self.k); // [1, 1000, 32]

        // 2. Gather neighbour features
        let points_exp = points
            .clone()
            .unsqueeze_dim::<4>(2)
            .repeat(&[1, 1, self.k, 1]); // [1, 1000, 32, 3]
        let idx_exp = idx.unsqueeze_dim::<4>(3).repeat(&[1, 1, 1, 3]); // [1, 1000, 32, 3]
        let neighbors = points_exp.gather(1, idx_exp); // [1, 1000, 32, 3]

        // 3. Edge features: h_θ(x_i) || h_φ(x_j - x_i)
        let x_i = points; // [1, 1000, 3]
        let x_j = neighbors; // [1, 1000, 32, 3]
        let diff = x_j - x_i.clone().unsqueeze_dim::<4>(2); // [1, 1000, 32, 3]

        let h_theta = self.theta.forward(x_i); // [1, 1000, 128]
        let h_phi = self.phi.forward(diff); // [1, 1000, 32, 128]

        let edge = h_theta.unsqueeze_dim::<4>(2) + h_phi; // [1, 1000, 32, 128]

        // Returning here works:
        // edge.sum_dim(2).squeeze::<3>(2)

        // 4. Max-pool over neighbours
        // This crashes: Unsupported dimension, only the last dimension can differ: Tensor [1, 1000, 32, 128] Index [1, 1000, 1, 128]
        let pooled = edge.max_dim(2).squeeze::<3>(2); // [1, 1000, 128]

        pooled
```

I have the same issue with `min_dim()` in my loss function. Note that there is no issue with `sum_dim()`, `mean_dim()`, `min()`, `max()`.

---

But at the same time, this works:
```rust
    pub fn forward(&self, points: Tensor<B, 3>) -> RegressionOutput<B> {
        let loss = points.clone().min_dim(2).squeeze::<2>(2).max_dim(1).squeeze::<1>(1).sum_dim(0);
        let output = Tensor::from_data(TensorData::zeros::<f32, _>(vec![1,0]), &points.device());
        let targets = output.clone();

        RegressionOutput::new(loss, output, targets)
    }
```

**Comment by laggui** [comment]
@awoimbee sorry for the delay, I just got around to investigating this today.

I just submitted a PR for the fix, thanks again for flagging this issue!
